src/math.c

```diff
@@ -189,137 +255,689 @@ u24_8 u24_8_div(u24_8 x, u24_8 y)
+            }
+            else { // r2 >= r3
+                divi = r2 / 256;
+                if (divi == 0)
+                    return 0xA00;
+
+                idx = r3 / divi;
+                if (idx > 0xFF)
+                    idx = 0xFF;
+            
+                return (0xC0 - gFastUnknownFn1Lookup[idx]) << 4;
+            }
+        }
+    }
+}
+
+static void F48_16_Negate(unkStruct_80943A8 *a)
+{
+    a->s0 = ~a->s0;
+    a->s4 = ~a->s4 + 1;
+
+    if (a->s4 == 0)
+        a->s0++;
+}
+
+static void F48_16_Abs(unkStruct_80943A8 *a)
+{
+    if (a->s0 < 0) {
+        a->s0 = ~a->s0;
+        a->s4 = ~a->s4 + 1;
+
+        if (a->s4 == 0)
+            a->s0++;
+    }
+}
+
+bool8 F48_16_IsZero(unkStruct_80943A8 *a)
+{
+    if (a->s0 == 0 && a->s4 == 0)
+        return TRUE;
+    return FALSE;
+}
+
+UNUSED bool8 F48_16_IsEqual(unkStruct_80943A8 *a, unkStruct_80943A8 *b)
+{
+    if (a->s0 == b->s0 && a->s4 == b->s4)
+        return TRUE;
+    return FALSE;
+}
+
+static bool8 F48_16_IsNegative(unkStruct_80943A8 *a)
+{
+    if (a->s0 < 0)
+        return TRUE;
+    return FALSE;
+}
+
+bool8 sub_800A2F0(unkStruct_80943A8 *a, unkStruct_80943A8 *b)
+{
+    s32 r1;
+    u32 a0;
+    s32 b0;
+
+    a0 = a->s0;
+    r1 = a0 >> 31;
+
+    b0 = b->s0;
+    if (b0 < 0)
+        r1 |= 0x2;
+
+    switch (r1) {
+        case 0:
+        default:
+            return u32_pair_less_than(a0, a->s4, b0, b->s4);
+        case 1:
+            return TRUE;
+        case 2:
+            return FALSE;
+        case 3:
+            return !u32_pair_less_than(a0, a->s4, b0, b->s4);
+    }
+}
+
+void F48_16_SMul(unkStruct_80943A8 *dst, unkStruct_80943A8 *a, unkStruct_80943A8 *b)
+{
+    bool8 aIsNegative;
+    bool8 bIsNegative;
+    unkStruct_80943A8 aa;
+    unkStruct_80943A8 bb;
+    unkStruct_80943A8 res;
+
+    aa.s0 = a->s0;
+    aa.s4 = a->s4;
+    bb.s0 = b->s0;
+    bb.s4 = b->s4;
+    aIsNegative = F48_16_IsNegative(&aa);
+    bIsNegative = F48_16_IsNegative(&bb);
+
+    if (F48_16_IsZero(&aa)) {
+        dst->s0 = 0;
+        dst->s4 = 0;
+    }
+    else if (F48_16_IsZero(&bb)) {
+        dst->s0 = 0;
+        dst->s4 = 0;
+    }
+    else {
+        if (aIsNegative)
+            F48_16_Negate(&aa);
+
+        if (bIsNegative)
+            F48_16_Negate(&bb);
+
+        F48_16_UMul(&res, &aa, &bb);
+        if (aIsNegative != bIsNegative)
+            F48_16_Negate(&res);
+
+        dst->s0 = res.s0;
+        dst->s4 = res.s4;
+    }
+}
+
+void F48_16_SDiv(unkStruct_80943A8 *dst, unkStruct_80943A8 *a, unkStruct_80943A8 *b)
+{
+    bool8 aIsNegative;
+    bool8 bIsNegative;
+    unkStruct_80943A8 aa;
+    unkStruct_80943A8 bb;
+    unkStruct_80943A8 res;
+
+    aa.s0 = a->s0;
+    aa.s4 = a->s4;
+    bb.s0 = b->s0;
+    bb.s4 = b->s4;
+    aIsNegative = F48_16_IsNegative(&aa);
+    bIsNegative = F48_16_IsNegative(&bb);
+
+    if (F48_16_IsZero(&bb)) {
+        dst->s0 = INT32_MAX;
+        dst->s4 = UINT32_MAX;
```

Comment:
I'm not sure using *INT32_MAX consts is semantically correct here, I think I'd prefer `~(1<<31)` (or `~0u>>1`) and `~0` actually